toqito/state_props/is_separable.py

```diff
@@ -66,247 +207,550 @@ def is_separable(state: np.ndarray, dim: None | int | list[int] = None, level: i
+            raise ValueError("The parameter `dim` must evenly divide the length of the state.")
+        else:
+            dims_list = [int(dim), int(np.round(state_len / dim))]
+    elif isinstance(dim, list) and len(dim) == 2:
+        if not all(isinstance(d, (int, np.integer)) and d >= 0 for d in dim):
+            raise ValueError("Dimensions in list must be non-negative integers.")
+        if dim[0] * dim[1] != state_len:
+            if (dim[0] == 0 or dim[1] == 0) and state_len != 0:
+                raise ValueError("Non-zero state with zero-dim subsystem is inconsistent.")
+            raise ValueError("Product of list dimensions must equal state length.")
+        dims_list = [int(d) for d in dim]
+    else:
+        raise ValueError("`dim` must be None, an int, or a list of two non-negative integers.")
+
+    dA, dB = dims_list[0], dims_list[1]
+    if (dA == 0 or dB == 0) and state_len != 0:
+        raise ValueError("Non-zero state with zero-dim subsystem is inconsistent.")
+
+    if state_len == 0:
+        return True
 
-    # Check the proximity of X with the maximally mixed state.
-    if in_separable_ball(state):
-        # Determined to be separable by closeness to the maximally mixed state. See (Gurvits_2002_Ball).
+    state_rank = np.linalg.matrix_rank(current_state, tol=tol)
+    min_dim_val, max_dim_val = min(dA, dB), max(dA, dB)
+    prod_dim_val = dA * dB
+
+    if prod_dim_val == 0 and state_len > 0:
+        raise ValueError("Zero product dimension for non-empty state is inconsistent.")
+    if prod_dim_val > 0 and prod_dim_val != state_len:
+        raise ValueError(f"Internal dimension calculation error: prod_dim {prod_dim_val} != state_len {state_len}")
+
+    # --- 2. Trivial Cases for Separability ---
+    if min_dim_val == 1:
+        # Every positive semidefinite matrix is separable when one of the local dimensions is 1.
         return True
 
-    # Check if X is a rank-1 perturbation of the identity, which is
-    # necessarily separable if it's PPT, which we have already checked.
-    if lam[1] - lam[prod_dim - 1] < tol**2:
-        # Determined to be separable by being a small rank-1 perturbation of the maximally-mixed state.
-        # See (Vidal_1999_Robust).
+    # --- 3. Pure State Check (Schmidt Rank) ---
+    # A pure state (rank 1) is separable if and only if its Schmidt rank is 1.
+    # (The condition `s_rank <= 2` was previously here, referencing Cariello for weak irreducible matrices;
+    # however, for general pure states, s_rank=1 is the N&S condition.
+    # TODO: Consider adding a separate check for OperatorSchmidtRank <= 2 for general mixed states
+    # if they are determined to be "weakly irreducible", as per Cariello :cite:`Cariello_2013_Weak_irreducible`
+    # and QETLAB's implementation. This is distinct from this pure state check.)
+    if state_rank == 1:
+        s_rank = schmidt_rank(current_state, dims_list)
+        return bool(s_rank == 1)
```

Comment:
Do we need to wrap this in a `bool`? It `s_rank == 1` should be a `bool` by default, right?